torch/csrc/jit/mobile/model_bytecode_version.cpp

```diff
@@ -0,0 +1,175 @@
+#include <ATen/core/ivalue.h>
+#include <caffe2/serialize/file_adapter.h>
+#include <caffe2/serialize/inline_container.h>
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/mobile/model_bytecode_version.h>
+#include <torch/csrc/jit/mobile/module.h>
+#include <torch/csrc/jit/serialization/import_read.h>
+#include <torch/csrc/jit/serialization/type_name_uniquer.h>
+#include <torch/custom_class.h>
+
+#include <exception>
+#include <string>
+#include <vector>
+
+namespace c10 {
+TypePtr parseType(const std::string& pythonStr);
+} // namespace c10
+
+namespace torch {
+namespace jit {
+
+using caffe2::serialize::FileAdapter;
+using caffe2::serialize::IStreamAdapter;
+using caffe2::serialize::PyTorchStreamReader;
+using caffe2::serialize::ReadAdapterInterface;
+
+namespace mobile {
+class CompilationUnit;
+}
+
+namespace {
+
+TypePtr resolveTypeName(
+    std::shared_ptr<CompilationUnit> compilation_unit,
+    const c10::QualifiedName& qn) {
+  // HACK: first we check whether the name starts with special prefix to
+  // tell if it's a supported pytorch class type. There are two special
+  // prefixes. "__torch__" for nn module, and "torch.jit" from to_backend.
+  // This is a reliable
+  // check today, but there is no guarantee that this is the case. The
+  // real solution is to merge type parsers so we can share class
+  // resolution logic.
+  static const c10::QualifiedName torchPrefix = "__torch__";
+  static const c10::QualifiedName jitPrefix = "torch.jit";
+  if (torchPrefix.isPrefixOf(qn) || jitPrefix.isPrefixOf(qn)) {
+    if (compilation_unit->get_class(qn) == nullptr) {
+      auto typeptr = ClassType::create(qn, compilation_unit, true);
+      compilation_unit->register_type(typeptr);
+    }
+    return compilation_unit->get_class(qn);
+  } else {
+    return c10::parseType(qn.qualifiedName());
+  }
+}
+
+c10::IValue readArchive(
```

Comment:
Are all these utilities copy pasted from somewhere else?

Reply:
Currently `mobile/import.cpp`, `jit/serialization/import.cpp` and this file uses the generic function `torch::jit::readArchiveAndTensors(archive_name, type_resolver, obj_loader, device, stream_reader)` to read archive, after #57098 lands. 

The `type_resolver` and `obj_loader` lambda functions here are the same as `mobile/import.cpp`, but different from `jit/serialization/import.cpp`. I can probably try separating them out from `mobile/import.cpp` and reuse it here.